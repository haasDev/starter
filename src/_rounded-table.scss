@mixin rounded-table($borderRadius, $borderColor, $borderWidth) {
  $halfRadius: ($borderRadius / 2);

  border: solid $borderColor $borderWidth;
  border-collapse: separate;
  border-radius: $borderRadius;
  border-spacing: 0;

  td,
  th {
    /* border: solid $borderColor $halfRadius; */
    border: solid $borderColor 1px;
    padding: $space;
  }

  th:first-child {
    border-top-left-radius: $tableRadius;
  }

  th:last-child {
    border-top-right-radius: $tableRadius;
  }

  tr:last-child td:first-child {
    border-bottom-left-radius: $tableRadius;
  }

  tr:last-child td:last-child {
    border-bottom-right-radius: $tableRadius;
  }

  td:first-child,
  th:first-child {
    border-left: none;
  }

  td:last-child,
  th:last-child {
    border-right: none;
  }

  tr:first-child th {
    border-top: none;
  }

  tr:last-child td {
    border-bottom: none;
  }
}
