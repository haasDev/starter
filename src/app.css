.data-display .delete-btn {
  background-color: transparent;
  border: none; }

.data-display table {
  border-collapse: collapse; }
  .data-display table caption {
    font-size: 18px;
    font-weight: bold;
    margin-top: 10px; }
  .data-display table td,
  .data-display table th {
    border: solid gray 1px;
    padding: 5px; }

.data-entry button {
  margin-left: 5px; }

.data-entry label {
  font-weight: bold;
  margin-right: 5px; }

.data-display a,
.data-entry a {
  display: block;
  margin-top: 10px; }

.error {
  color: red;
  font-weight: bold; }
