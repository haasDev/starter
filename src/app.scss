.data-display {
  table {
    border-collapse: collapse;

    caption {
      font-size: 18px;
      font-weight: bold;
      margin-top: 10px;
    }

    td,
    th {
      border: solid gray 1px;
      padding: 5px;
    }
  }
}

.data-entry {
  a {
    display: block;
    margin-top: 10px;
  }

  label {
    font-weight: bold;
    margin-right: 5px;
  }
}
