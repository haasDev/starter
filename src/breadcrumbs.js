import React, {PropTypes as t} from 'react'
import {Breadcrumb} from 'react-bootstrap'

const {Item} = Breadcrumb

const Breadcrumbs = ({activeCrumb, items, onNavigate}) => {
  const activeId = activeCrumb ?
    activeCrumb.id :
    items.length ? items[items.length - 1].id : null
  return (
    <Breadcrumb>
      {items.map(item => {
        const active = activeId === item.id
        const handlers = active ? {} : {onClick: () => onNavigate(item)}
        return (
          <Item active={active} key={item.id} {...handlers}>
            {item.label}
          </Item>
        )
      })}
    </Breadcrumb>
  )
}

Breadcrumbs.displayName = 'Breadcrumbs'

Breadcrumbs.propTypes = {
  activeCrumb: t.shape({
    id: t.number,
    label: t.string,
  }),
  items: t.arrayOf(
    t.shape({
      id: t.number.isRequired,
      label: t.string.isRequired,
    }),
  ).isRequired,
  onNavigate: t.func.isRequired,
}

export default Breadcrumbs
