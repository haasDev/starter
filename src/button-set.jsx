// @flow
import React, {PropTypes as t} from 'react';

// Types
type ButtonKindType =
  | 'danger'
  | 'default'
  | 'info'
  | 'primary'
  | 'secondary'
  | 'success';

type ButtonType = {
  text: string,
  kind: ButtonKindType,
  onClick: () => any
};

type ButtonSetPropType = {
  buttons: Array<ButtonType>
};

// View
const ButtonSet = ({buttons}: ButtonSetPropType) => (
  <div>
    {buttons.map(button => (
      <button
        className={`btn btn-${button.kind}`}
        key={button.text}
        onClick={button.onClick}
      >
        {button.text}
      </button>
    ))}
  </div>
);

// PropTypes
ButtonSet.propTypes = {
  buttons: t.arrayOf(
    t.shape({
      text: t.string,
      kind: t.oneOf([
        'danger',
        'default',
        'info',
        'primary',
        'secondary',
        'success',
      ]),
      onClick: t.func,
    }),
  ).isRequired,
};

export default ButtonSet;

