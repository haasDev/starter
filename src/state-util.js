export let setState;

export function defineSetState(component) {
  setState = component.setState.bind(component);
}
