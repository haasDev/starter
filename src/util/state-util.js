export let setState;

/**
 * This sets `setState` to a function that
 * components use to change the app state.
 * Typically the top component will be passed
 * to this so its state will be used.
 * For tests it is useful to pass a function
 * so assertions about state changes can be made.
 */
export function defineSetState(thing) {
  setState = typeof thing === 'function' ? thing : thing.setState.bind(thing);
}
