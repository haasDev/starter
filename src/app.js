// @flow
import React, {Component} from 'react';
import Breadcrumbs from './breadcrumbs';
import ButtonSet from './button-set';
import DataEntry from './data-entry';
import DataDisplay from './data-display';
import DropupBtn from './dropup-button';
import LookupInput from './lookup-input';
import {defineSetState, setState} from './state-util';
import {getLocationParts} from './hash-route';
import {getUrl} from './url-util';
import {handleError} from './error';
import './app.css';

type BreadcrumbType = {
  id: number,
  label: string
};

async function loadProjects() {
  const url = getUrl('project');
  try {
    const res = await fetch(url);
    if (!res.ok) return handleError(url, res);

    const projects = await res.json();

    const projectMap = {};
    for (const project of projects) {
      projectMap[project.id] = project;
    }

    setState({projectMap});
  } catch (e) {
    handleError(url, e);
  }
}

class App extends Component {
  state = {
    activeCrumb: undefined,
    description: '',
    error: '',
    name: '',
    projectMap: {},
  };

  breadcrumbs = [
    {id: 1, label: 'Foo'},
    {id: 2, label: 'Bar'},
    {id: 3, label: 'Baz'},
  ];

  constructor() {
    super();

    // Allow any component to change the state of this top-most component.
    defineSetState(this);

    // Re-render any time the URL hash changes.
    window.addEventListener('hashchange', () => this.forceUpdate());
  }

  componentDidMount() {
    loadProjects();
  }

  onNavigate = (crumb: BreadcrumbType) => {
    this.setState({activeCrumb: crumb});
  };

  render() {
    const {hash} = getLocationParts(window.location);
    const {activeCrumb, description, error, name, projectMap} = this.state;
    const buttons = [
      {
        disabled: true,
        text: 'Save',
        kind: 'primary',
        onClick: () => console.log('saved!'),
      },
      {
        disabled: false,
        text: 'Info',
        kind: 'info',
        onClick: () => console.log('info!'),
      },
      {
        text: 'Cancel',
        kind: 'danger',
        onClick: () => console.log('cancelled!'),
      },
    ];

    const input = {
      img: 'search',
      onChange: (event = {}) => console.log(event.target.value),
      onSubmit: () => console.log('clicked!'),
    };

    const dropupBtnParams = {
      btn: {
        disabled: false,
        kind: 'danger',
        btnText: 'My button',
      },
      links: [
        {
          onClick: () => console.log('clicked1'),
          separator: false,
          text: 'link 1',
        },
        {
          onClick: () => console.log('clicked2'),
          separator: false,
          text: 'link 2',
        }
      ],
    };

    return (
      <div className="app">
        <Breadcrumbs
          activeCrumb={activeCrumb}
          items={this.breadcrumbs}
          onNavigate={this.onNavigate}
        />
        <div className="error">{error}</div>
        <div className="body">
          {hash === 'display' ?
            <DataDisplay projectMap={projectMap} /> :
            <DataEntry description={description} name={name} />}
          <ButtonSet buttons={buttons} />
          <LookupInput {...input} />
          <DropupBtn {...dropupBtnParams} />
        </div>
      </div>
    );
  }
}

export default App;
