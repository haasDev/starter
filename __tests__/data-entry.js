import React from 'react';
import DataEntry from '../src/data-entry';
import {defineSetState} from '../src/state-util';
import {mount} from 'enzyme';
import snapshot from './snapshot';

describe('DataEntry', () => {
  const projectMap = {
    1: {id: 1, name: 'Foo', description: 'desc 1'},
    2: {id: 2, name: 'Bar', description: 'desc 2'}
  };

  let state;

  function addCheck(state, projectDescription, projectName, done) {
    setTimeout(() => { // run in next tick
      try {
        expect(state.name).toBe('');
        const id = 3;
        const project = state.projectMap[id];
        expect(project).toEqual(
          {id, name: projectName, description: projectDescription});
        done();
      } catch (e) {
        done.fail(e);
      }
    });
  }

  function addSetup() {
    // Mock the fetch call to add a project, returning id 3.
    window.fetch = jest.fn().mockImplementation((url, options) => {
      expect(options.method).toBe('POST');
      return Promise.resolve({
        ok: true,
        text: () => 3 // id of new project
      });
    });

    defineSetState(thing => {
      const newState = typeof thing === 'function' ? thing(state) : thing;
      state = Object.assign(state, newState);
    });
  }

  it('should render', () => {
    const name = 'My Project';
    const description = 'My description';
    snapshot(<DataEntry description={description} name={name} />);
  });

  it('should support name entry', () =>
    new Promise(resolve => {
      const expectedName = 'My New Project';
      defineSetState(newState => {
        expect(newState).toEqual({name: expectedName});
        resolve();
      });

      const wrapper = mount(<DataEntry description="" name="" />);
      const input = wrapper.find('input').first();
      input.simulate('change', {target: {value: expectedName}});
    }));

  it('should add a project by pressing Add button', done => {
    addSetup();

    state = {name: '', description: '', projectMap};
    const onKeyPress = jest.fn();
    const projectName = 'My New Project';
    const projectDescription = 'My Project desc';
    const wrapper = mount(
      <DataEntry name="name" value={projectName} onKeyPress={onKeyPress} />
    );
    const addBtn = wrapper.find('.add-btn');
    addBtn.simulate('click');

    addCheck(state, projectDescription, projectName, done);
  });

  it('should add a project by pressing enter key', done => {
    addSetup();

    state = {name: '', projectMap};
    const projectName = 'My New Project';
    const description = 'My description';
    const wrapper = mount(
      <DataEntry description={description} name={projectName} />
    );
    const input = wrapper.find('input');
    input.simulate('keyPress', {which: 13});

    addCheck(state, projectName, done);
  });
});
