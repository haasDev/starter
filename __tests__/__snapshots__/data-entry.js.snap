exports[`DataEntry should render 1`] = `
<div
  className="data-entry">
  <div>
    <span>
      <label>
        Project Name
      </label>
      <input
        autoFocus={undefined}
        name="name"
        onChange={[Function]}
        onKeyPress={undefined}
        type="text"
        value="My Project" />
    </span>
    <span>
      <label>
        Project Description
      </label>
      <input
        autoFocus={undefined}
        name="description"
        onChange={[Function]}
        onKeyPress={[Function]}
        type="text"
        value={undefined} />
    </span>
    <button
      className="add-btn"
      onClick={[Function]}>
      Add
    </button>
  </div>
  <a
    href="#display">
    Show Projects
  </a>
</div>
`;
